resources/testharness.js

```diff
@@ -2016,19 +2138,29 @@ policies and contribution forms [3].
         if (this.phase === this.phases.COMPLETE) {
             return;
         }
-        this.phase = this.phases.COMPLETE;
         var this_obj = this;
-        this.tests.forEach(
-            function(x)
-            {
-                if (x.phase < x.phases.COMPLETE) {
-                    this_obj.notify_result(x);
-                    x.cleanup();
-                    x.phase = x.phases.COMPLETE;
-                }
-            }
-        );
-        this.notify_complete();
+
+        all_async(this.tests,
+                  function(test, testDone)
+                  {
+                      if (test.phase === test.phases.COMPLETE) {
+                          testDone();
+                          return;
+                      }
+
+                      test._add_done_callback(testDone);
+
+                      if (test.phase < test.phases.CLEANING) {
+                          test.cleanup();
```

Comment:
I see this was done before as well, but can you comment about why cleanup is started here, when in some cases apparently it can already have happened?

Reply:
As near as I can tell, this method was originally written to invoke the
`cleanup` method as a way to simulate completion for pending tests. But that's
pretty indirect, as your question indicates. It would be better to do something
like the following:

```js
if (test.phase === test.phases.CLEANING) {
    test._add_done_callback(testDone);
    return;
}

if (test.phase < test.phases.CLEANING) {
    test.phase = test.phases.COMPLETE;
    this_obj.result(test);
}

testDone();
```

...but this version causes failures, so I will need to investigate further.
I'll have a better answer for you tomorrow.

Reply:
Leaving a comment or a TODO would be fine, given that this was already odd.

Reply:
I was able to address this as part of the refactoring mentioned above.